In [62]:
"""
!pip install shap
!pip install xgboost
!pip install sklearn
!pip install category_encoders
!pip install pymongo
"""

In [1]:
import re
import time
import matplotlib.pyplot as plt
from pymongo import MongoClient
import pandas as pd
import numpy as np
import warnings
import random
import sqlite3
warnings.filterwarnings('ignore')

In [2]:
username = 'kwanhoon'
password = 'shk9293112'

URI = f'mongodb+srv://{username}:{password}@cluster0.xsktws7.mongodb.net/A1R?retryWrites=true&w=majority'

client = MongoClient(URI)  # 파이썬과 mongo를 연결
DATABASE = 'IT_product'

database = client[DATABASE]

COLLECTION = 'Monitor'
COLLECTION_32 = 'Monitor_32'
collection_27 = database[COLLECTION]
collection_32 = database[COLLECTION_32]

list_27 = []
list_32 = []

for d in collection_32.find():
    list_32.append(d)
    
for d in collection_27.find():
    list_27.append(d)

dict_27 = list_27[0]
dict_32 = list_32[0]

In [3]:
df_27 = pd.DataFrame.from_dict(dict_27).T.iloc[1:, 0:9]
df_32 = pd.DataFrame.from_dict(dict_32).T.iloc[1:, 0:9]


display(df_27.head(3))
df_32.head(3)

화면크기             최대 해상도 최대 주사율  \
인터픽셀 IP2722 75                  28인치(71cm)  3840x2160(4K UHD)   60Hz   
인터픽셀 IP2740 게이밍 165          27인치(68~69cm)     1920x1080(FHD)  144Hz   
한성컴퓨터 ULTRON 2768 BLADE New     28인치(71cm)  3840x2160(4K UHD)  144Hz   

                                      패널  곡면형  화면비율 화면종류 응답속도  \
인터픽셀 IP2722 75               IPS패널(광시야각)   평면  16:9  LED  5ms   
인터픽셀 IP2740 게이밍 165            V (광시야각)A  NaN  16:9  LED  1ms   
한성컴퓨터 ULTRON 2768 BLADE New          IPS   평면  16:9  LED  1ms   

                                              부가기능  
인터픽셀 IP2722 75                                 NaN  
인터픽셀 IP2740 게이밍 165          HDCP지원, , , , , , , ,  
한성컴퓨터 ULTRON 2768 BLADE New                    NaN

화면크기                 최대 해상도 최대 주사율         패널  \
인터픽셀 IPQ3245 165     32인치(80~81cm)         2560x1440(QHD)  144Hz  VA( 광시야각)   
크로스오버존 크로스오버 32ULP3  32인치(80~81cm)      3840x2160(4K UHD)   60Hz        VAA   
인터픽셀 IPU3433         34인치(86~87cm)  3440x1440(Ultra WQHD)  165Hz        IPS   

                     화면비율 화면종류 응답속도                         부가기능  곡면형  
인터픽셀 IPQ3245 165     16:9  LED  1ms                       HDCP지원  NaN  
크로스오버존 크로스오버 32ULP3  16:9  LED  4ms  스피커 내장, 웹캠내장, , , , , , , ,   평면  
인터픽셀 IPU3433         21:9  NaN  1ms                          NaN  NaN

### 결측값 처리

* 화면 최대 해상도 -> 대부분 FHD
* 최대 주사율 60 or 144 or 165?
* 화면 비율 -> 칼럼 삭제
* 화면 종류 -> 칼럼 삭제
* 부가기능 -> 칼럼 삭제
* 응답 속도 ->  최대주사율 60일때 5, 144이상일 때, 1
* 곡면형 -> 최빈값

### 칼럼값 정제

* 최대 해상도 -> UHD, FHD, QHD로 크게 3개로 나눔
* 패널 -> ips, tn, va로 크게 3개로 나눔

In [4]:
df_27['화면크기'] = ['27'] * df_27.shape[0]
df_32['화면크기'] = ['32'] * df_32.shape[0]

def feature_engineer(df):
    # 칼럼 삭제
    del_col = ['화면비율', '부가기능', '화면종류']
    df.drop(del_col, axis = 1, inplace = True)
    
    # 곡면형 칼럼의 평면, , ,를 평면으로 변경
    idx = df[df['곡면형'].str.contains(',') == True].index
    
    df.loc[idx, '곡면형'] = df[df['곡면형'].str.contains(',') == True].loc[:, '곡면형'].str.split(',').str[0]
    df['곡면형'].fillna(df['곡면형'].mode()[0], inplace = True)
    
    # 최대 해상도에서 갯수가 10개 이하인 칼럼의 데이터는 삭제
    idx_pixel = df['최대 해상도'].value_counts()[df['최대 해상도'].value_counts() < 10].index
    for i in idx_pixel:
        df = df.drop(df[df['최대 해상도'] == i].index, axis = 0)
    
    idx_FHD = list(df[df['최대 해상도'].str.contains('FHD') == True]['최대 해상도'].index)
    df.loc[idx_FHD, '최대 해상도'] = ['FHD'] * len(idx_FHD)
    
    idx_QHD = list(df[df['최대 해상도'].str.contains('QHD') == True]['최대 해상도'].index)
    df.loc[idx_QHD, '최대 해상도'] = ['QHD'] * len(idx_QHD)
    
    idx_UHD = list(df[df['최대 해상도'].str.contains('UHD') == True]['최대 해상도'].index)
    df.loc[idx_UHD, '최대 해상도'] = ['UHD'] * len(idx_UHD)
    
    df['최대 해상도'].fillna(df['최대 해상도'].mode()[0], inplace = True)
    
    # 최대 주사율에서 Hz 빼기
    ## 최대 주사율 결측값은 최빈값으로 대체
    df['최대 주사율'] = df['최대 주사율'].str.replace(r'\D', '')
    df['최대 주사율'].fillna(df['최대 주사율'].mode()[0], inplace = True)
    
    df.loc[:, '최대 주사율'] = df.loc[:, '최대 주사율'].astype(int)
    
    
    # 패널 종류
    idx_ips = (df[df['패널'].str.contains('IPS|ips') == True]['패널'].index)
    df.loc[idx_ips, '패널'] = ['IPS'] * len(idx_ips)
    
    idx_va = (df[df['패널'].str.contains('VA|va') == True]['패널'].index)
    df.loc[idx_va, '패널'] = ['VA'] * len(idx_va)
    
    idx_tn = (df[df['패널'].str.contains('TN|tn') == True]['패널'].index)
    df.loc[idx_tn, '패널'] = ['TN'] * len(idx_tn)
    
    idx_I = (df[df['패널'].str.contains('I') == True]['패널'].index)
    df.loc[idx_I, '패널'] = ['IPS'] * len(idx_I)
    
    idx_V = (df[df['패널'].str.contains('V') == True]['패널'].index)
    df.loc[idx_V, '패널'] = ['VA'] * len(idx_V)
    
    idx_panel = df['패널'].value_counts()[df['패널'].value_counts() < 30].index
    for i in idx_panel:
        df = df.drop(df[df['패널'] == i].index, axis = 0)
    
    df['패널'].fillna(df['패널'].mode()[0], inplace = True)
    
    # 응답속도
    df['응답속도'] = df['응답속도'].str.replace(r'\D', '')
    
    pixel = ['FHD', 'QHD', 'UHD']
    
    for p in pixel:
        mode_p = df[df['최대 해상도'] == p]['응답속도'].mode()[0]
        idx_p = list((df['최대 해상도'] == p).index)
        df.loc[idx_p , '응답속도'] = df.loc[idx_p, '응답속도'].fillna(mode_p)

    df.loc[:, '응답속도'] = df.loc[:, '응답속도'].astype(float)
    
    idx_speed = list((df['응답속도']>=10).index)
    df.loc[idx_speed, '응답속도'] = df.loc[idx_speed, '응답속도']/10
    
    star = []
    for i in range(df.shape[0]):
        star.append(round(random.uniform(3.5, 5), 2))
    
    df['별점'] = star

    return df

In [5]:
df_27_pre = feature_engineer(df_27.copy())
df_32_pre = feature_engineer(df_32.copy())

display(df_27_pre.info())
display(df_32_pre.info())

df = pd.concat([df_27_pre, df_32_pre], axis = 0).reset_index().rename(columns = {'index' : '상품명'})
#df = df.reset_index().rename(columns = {'index' : '상품명'})

<class 'pandas.core.frame.DataFrame'>
Index: 2984 entries, 인터픽셀 IP2722 75 to 3크로스오버 2890CU IPS UHD HDR TYPE-C 평면
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   화면크기    2984 non-null   object 
 1   최대 해상도  2984 non-null   object 
 2   최대 주사율  2984 non-null   int64  
 3   패널      2984 non-null   object 
 4   곡면형     2984 non-null   object 
 5   응답속도    2984 non-null   float64
 6   별점      2984 non-null   float64
dtypes: float64(2), int64(1), object(4)
memory usage: 251.0+ KB


None

<class 'pandas.core.frame.DataFrame'>
Index: 2132 entries, 인터픽셀 IPQ3245 165 to LG전자 LG전자 80cm 32인치 FHD IPS 모니터 32MN600PW
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   화면크기    2132 non-null   object 
 1   최대 해상도  2132 non-null   object 
 2   최대 주사율  2132 non-null   int64  
 3   패널      2132 non-null   object 
 4   응답속도    2132 non-null   float64
 5   곡면형     2132 non-null   object 
 6   별점      2132 non-null   float64
dtypes: float64(2), int64(1), object(4)
memory usage: 197.8+ KB


None

In [6]:
"""
con = sqlite3.connect('/Users/kwanhoonseo/Desktop/sc3_pj/monitor.db')
df.to_sql('Monitor', con)
"""

"\ncon = sqlite3.connect('/Users/kwanhoonseo/Desktop/sc3_pj/monitor.db')\ndf.to_sql('Monitor', con)\n"

In [8]:
import psycopg2

conn = psycopg2.connect(
    host = "peanut.db.elephantsql.com",
    database = "pblabcdk",
    user = "pblabcdk",
    password = "HPGjhgcQnKL8_30d3RqYF574gmNSWDUj"
)

cur = conn.cursor()

In [10]:
df.to_sql('Monitor', conn)

DatabaseError: Execution failed on sql 'SELECT name FROM sqlite_master WHERE type='table' AND name=?;': relation "sqlite_master" does not exist
LINE 1: SELECT name FROM sqlite_master WHERE type='table' AND name=?...
                         ^
